src/librekey/key_store_g10.c

```diff
@@ -441,13 +527,305 @@ parse_seckey(pgp_keydata_key_t *keydata, s_exp_t *s_exp, pgp_pubkey_alg_t alg)
     return true;
 }
 
+static pgp_seckey_t *
+g10_decrypt_seckey(const pgp_key_t *key, FILE *passfp)
+{
+    uint8_t        derived_key[PGP_MAX_KEY_SIZE];
+    char *         passphrase;
+    char           pass[MAX_PASSPHRASE_LENGTH];
+    unsigned       keysize;
+    uint8_t *      decrypted;
+    pgp_seckey_t * seckey;
+    s_exp_t        s_exp = {0};
+    size_t         output_written = 0;
+    size_t         input_consumed = 0;
+    botan_cipher_t decrypt;
+
+    if (key->key.seckey.encrypted_len == 0) {
+        fprintf(stderr, "Hasn't got encrypted data!\n");
+        return NULL;
+    }
+
+    if (pgp_getpassphrase(passfp, pass, sizeof(pass)) == 0) {
+        pass[0] = '\0';
+    }
+
+    passphrase = rnp_strdup(pass);
+    pgp_forget(pass, sizeof(pass));
+    if (passphrase == NULL) {
+        (void) fprintf(stderr, "bad allocation\n");
+        return false;
+    }
+
+    keysize = pgp_key_size(key->key.seckey.alg);
+    if (keysize == 0) {
+        (void) fprintf(stderr, "parse_seckey: unknown symmetric algo\n");
+        pgp_forget(passphrase, strlen(passphrase));
+        free(passphrase);
+        return false;
+    }
+
+    if (pgp_s2k_iterated(key->key.seckey.hash_alg,
+                         derived_key,
+                         keysize,
+                         passphrase,
+                         key->key.seckey.salt,
+                         key->key.seckey.s2k_iterations)) {
+        (void) fprintf(stderr, "pgp_s2k_iterated failed\n");
+        pgp_forget(passphrase, strlen(passphrase));
+        free(passphrase);
+        return false;
+    }
+
+    pgp_forget(passphrase, strlen(passphrase));
+    free(passphrase);
+
+    if (rnp_get_debug(__FILE__)) {
+        hexdump(stderr, "input iv", key->key.seckey.iv, G10_CBC_IV_SIZE);
+        hexdump(stderr, "key", derived_key, keysize);
+        hexdump(stderr, "encrypted", key->key.seckey.encrypted, key->key.seckey.encrypted_len);
+    }
+
+    decrypted = malloc(key->key.seckey.encrypted_len);
+    if (decrypted == NULL) {
+        (void) fprintf(stderr, "can't allocate memory\n");
+        return false;
+    }
+
+    switch (key->key.seckey.cipher_mode) {
```

Comment:
Should the formats table be used here? It seems to assume AES-128 (no 256) below as well.